### PR TITLE
[Merged by Bors] - feat(GroupTheory/GroupAction/Blocks): lemmas to handle blocks on finite sets

### PR DESCRIPTION
**Pull request by AntoineChambert-Loir**
This PR proves specific properties of blocks when the action is on a finite set

---

- [x] depends on: #12023 
- [x] depends on: #12049 

It is a repost of PR #12051 which has been corrupted by a wrong merge.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (jcommelin) [inline Mathlib/GroupTheory/GroupAction/Blocks.lean:419]: Please structure proofs with bullet points.

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (jcommelin) [inline Mathlib/GroupTheory/GroupAction/Blocks.lean:453]: To enable dot notation:
```suggestion
theorem IsBlock.eq_top_of_large [hfX : Finite X] [hGX : IsPretransitive G X] {B : Set X}
```

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (jcommelin) [inline Mathlib/GroupTheory/GroupAction/Blocks.lean:432]: It seems that many of the following lemmas have these assumptions. Would it make sense to start a
```
namespace IsBlock

variable [Finite X] [IsPretransitive G X] {B : Set X}
```
and then also remove the `IsBlock` from the names?

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/GroupAction/Blocks.lean:453]: I am not sure I understood what I should do…

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/GroupAction/Blocks.lean:432]: I did that.

Comment (semorrison) [reply Mathlib/GroupTheory/GroupAction/Blocks.lean]: Please resolve conversations when you've made the changes, to make it easier for later reviewers to see the state of the PR.

Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/GroupAction/Blocks.lean]: OK. I was waiting for Johann's approval.

**Review by YaelDillies — COMMENTED** [review]
This all feels like it should be much more direct from orbit-stabilizer-like theorems. Maybe @tb65536 has thoughts here?

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/GroupAction/Blocks.lean]: Yes, but #find_home! did not find any which was appropriate.
I put the second lemma in Mathlib/Data/Set
For the first one, any place I thought of required to add some import and I have no clue about which place is best.

File: Mathlib/Data/Set/Card.lean
Comment (tb65536) [inline Mathlib/Data/Set/Card.lean:729]: Is this lemma used anywhere in the PR? If not, you could move it to a separate PR.

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (tb65536) [inline Mathlib/GroupTheory/GroupAction/Blocks.lean:65]: Is this lemma used anywhere in the PR? If not, you could move it to a separate PR.

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (tb65536) [inline Mathlib/GroupTheory/GroupAction/Blocks.lean:512]: The first three theorems do not require `[Finite X]`.

File: Mathlib/Data/Set/Card.lean
Comment (AntoineChambert-Loir) [reply Mathlib/Data/Set/Card.lean:729]: You're right. I deleted them.

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/GroupAction/Blocks.lean:65]: In fact, I don't use it anymore. Thanks for noting.

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/GroupAction/Blocks.lean:512]: Good point. I deleted the hypothesis and only add it for the lemmas that need it.

File: Mathlib/GroupTheory/Index.lean
Comment (tb65536) [inline Mathlib/GroupTheory/Index.lean:473]: This line is extra.

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (tb65536) [inline Mathlib/GroupTheory/GroupAction/Blocks.lean:544]: I think it is necessary. If you start the proof with the reversible steps
```
  rw [mk_subset]
  intro g x h1 h2 y h3
  obtain ⟨gx, rfl⟩ := exists_smul_eq G a x
  obtain ⟨gy, rfl⟩ := exists_smul_eq G a y
  simp only [Set.mem_smul_set_iff_inv_smul_mem, Set.mem_iInter] at h1 h2 h3 ⊢
```
then you get the goal state
```
B : Set X
a : X
g gx gy : G
h1 : ∀ (i : G), i⁻¹ • a ∈ B → i⁻¹ • gx • a ∈ B
h2 : ∀ (i : G), i⁻¹ • a ∈ B → i⁻¹ • g⁻¹ • gx • a ∈ B
h3 : ∀ (i : G), i⁻¹ • a ∈ B → i⁻¹ • g⁻¹ • gy • a ∈ B
⊢ ∀ (i : G), i⁻¹ • a ∈ B → i⁻¹ • gy • a ∈ B
```
which cannot be proven without further assumptions: `h1` lets you right-multiply by `gx`, `h2` lets you right-multiply by `g⁻¹ * gx`, `h3` lets you right-multiply by `g⁻¹ * gy`, but together these do not let you right-multiply by `gy`. You need some sort of finiteness or periodicity so that you can (for instance) right multiply by `gx * (g⁻¹ * gx)⁻¹ * (g⁻¹ * gy)`.

I think a counterexample might be `G = ℤ` acting on `ℤ` by translation, `B = ℕ`, and `a = 0`.

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/GroupAction/Blocks.lean:203]: Please add the deprecated aliases

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/GroupAction/Blocks.lean:567]: You could also write:
```suggestion
  have hag' (g : G) (hg : a ∈ g • B') : B' = g • B' := by
```

File: Mathlib/GroupTheory/Index.lean
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Index.lean:566]: I think you can use `variable` here

File: Mathlib/GroupTheory/Index.lean
Comment (Ruben-VandeVelde) [inline Mathlib/GroupTheory/Index.lean:571]: There's recently been some discussion against primes in names - can you try thinking of another way to distinguish these two?

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/GroupAction/Blocks.lean:492]: bot fix style

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/GroupAction/Blocks.lean:203]: Why not, but nobody ever imported that file, so I do not really see the point of adding a deprecated alias.

File: Mathlib/GroupTheory/Index.lean
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/Index.lean]: good point, I renamed it `index_stabilizer_of_transitive`

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/GroupAction/Blocks.lean:203]: Anyway, I did it. (Maybe that will help me…)

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (tb65536) [inline Mathlib/GroupTheory/GroupAction/Blocks.lean:206]: Do you mean to say `alias isFixedBlock_top := isFixedBlock_univ`?

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (tb65536) [inline Mathlib/GroupTheory/GroupAction/Blocks.lean:503]:
```suggestion
/-- The cardinality of the ambient space is the product of the cardinality of a block
  by the cardinality of the set of translates of that block -/
```

**Review by tb65536 — APPROVED** [review]

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (AntoineChambert-Loir) [reply Mathlib/GroupTheory/GroupAction/Blocks.lean:206]: Yes, and it was not at the right place.

File: Mathlib/GroupTheory/GroupAction/Blocks.lean
Comment (tb65536) [inline Mathlib/GroupTheory/GroupAction/Blocks.lean:501]:
```suggestion
/-- The cardinality of the ambient space is the product of the cardinality of a block
```

**Review by tb65536 — APPROVED** [review]

**Comment by tb65536** [comment]
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

File: Mathlib/GroupTheory/Index.lean
Comment (tb65536) [inline Mathlib/GroupTheory/Index.lean:9]:
```suggestion
```

**Comment by tb65536** [comment]
maintainer merge

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**
